File: language/runtime/apple/core/Sources/RuntimeHostAppleCore/Input/Text/TextInputRequests.swift
import Foundation

/// The text-input request surface attached to one Apple runtime host.
@MainActor
public protocol TextInputRequests {
  /// Open one Apple text-input session.
  func openTextInput(
    _ request: RuntimeHostTextInputOpenRequest
  ) -> UInt32

  /// Close one Apple text-input session.
  func closeTextInput(
    _ request: RuntimeHostTextInputCloseRequest
  ) -> UInt32

  /// Update one Apple text-input geometry hint.
  func setTextInputGeometry(
    _ request: RuntimeHostTextInputGeometryRequest
  ) -> UInt32

  /// Update one Apple text-input state.
  func setTextInputState(
    _ request: RuntimeHostTextInputStateRequest
  ) -> UInt32
}

/// The explicit unsupported text-input request surface for one Apple runtime host.
@MainActor
public final class UnsupportedTextInputRequests: TextInputRequests {
  /// Create one unsupported text-input request surface.
  public init() {}

  public func openTextInput(
    _ request: RuntimeHostTextInputOpenRequest
  ) -> UInt32 {
    let _ = request

    return hostStatusNotSupported
  }

  public func closeTextInput(
    _ request: RuntimeHostTextInputCloseRequest
  ) -> UInt32 {
    let _ = request

    return hostStatusNotSupported
  }

  public func setTextInputGeometry(
    _ request: RuntimeHostTextInputGeometryRequest
  ) -> UInt32 {
    let _ = request

    return hostStatusNotSupported
  }

  public func setTextInputState(
    _ request: RuntimeHostTextInputStateRequest
  ) -> UInt32 {
    let _ = request

    return hostStatusNotSupported
  }
}

File: language/runtime/android/kotlin/src/main/kotlin/dev/destack/runtime/android/input/text/TextInputRequests.kt
package dev.destack.runtime.android.input.text

import dev.destack.runtime.android.core.hostStatusNotSupported

/**
 * The text-input request surface attached to one Android runtime host.
 */
public interface TextInputRequests {
    /**
     * Open one Android text-input session.
     */
    public fun openTextInput(
        request: RuntimeHostTextInputOpenRequest,
    ): Int

    /**
     * Close one Android text-input session.
     */
    public fun closeTextInput(
        request: RuntimeHostTextInputCloseRequest,
    ): Int

    /**
     * Update one Android text-input geometry hint.
     */
    public fun setTextInputGeometry(
        request: RuntimeHostTextInputGeometryRequest,
    ): Int

    /**
     * Update one Android text-input state.
     */
    public fun setTextInputState(
        request: RuntimeHostTextInputStateRequest,
    ): Int
}

/**
 * The explicit unsupported text-input request surface for one Android runtime host.
 */
public object UnsupportedTextInputRequests : TextInputRequests {
    override fun openTextInput(
        request: RuntimeHostTextInputOpenRequest,
    ): Int {
        return hostStatusNotSupported
    }

    override fun closeTextInput(
        request: RuntimeHostTextInputCloseRequest,
    ): Int {
        return hostStatusNotSupported
    }

    override fun setTextInputGeometry(
        request: RuntimeHostTextInputGeometryRequest,
    ): Int {
        return hostStatusNotSupported
    }

    override fun setTextInputState(
        request: RuntimeHostTextInputStateRequest,
    ): Int {
        return hostStatusNotSupported
    }
}
